corehq/apps/app_manager/util.py

```diff
@@ -409,46 +409,83 @@ def _add_ancestors(ctype, generation):
+        "name": "{} ({})".format(generation_names[i], " or ".join(ctypes)) if i > 0 else base_case_type,
+        "key": "@case_type",
+        "structure": {p: {} for type in [map[t] for t in ctypes] for p in type},
+        "related": {"parent": {
+            "hashtag": "#case/" + generation_names[i + 1],
+            "subset": generation_names[i + 1],
+            "key": "@case_id",
+        }} if i < len(generations) - 1 else None,
+    } for i, ctypes in enumerate(generations)]
+
+    if is_usercase_in_use(app.domain):
+        subsets.append({
+            "id": USERCASE_TYPE,
+            "name": "user",
+            "key": "@case_type",
+            "structure": {p: {} for p in get_usercase_properties(app)},
+        })
+
     return {
         "id": "casedb",
         "uri": "jr://instance/casedb",
         "name": "case",
         "path": "/casedb/case",
         "structure": {},
-        "subsets": [{
-            "id": generation_names[i],
-            "name": "{} ({})".format(generation_names[i], " or ".join(ctypes)) if i > 0 else base_case_type,
-            "key": "@case_type",
-            "structure": {p: {} for type in [map[t] for t in ctypes] for p in type},
-            "related": {"parent": generation_names[i + 1]} if i < len(generations) - 1 else None,
-        } for i, ctypes in enumerate(generations)],
+        "subsets": subsets,
     }
 
 
 def get_session_schema(form):
     """Get form session schema definition
     """
     from corehq.apps.app_manager.suite_xml.sections.entries import EntriesHelper
+    app = form.get_app()
     structure = {}
-    datums = EntriesHelper(form.get_app()).get_datums_meta_for_form_generic(form)
+    datums = EntriesHelper(app).get_datums_meta_for_form_generic(form)
     datums = [
         d for d in datums
         if not d.is_new_case_id and d.case_type and d.requires_selection
     ]
     if len(datums):
         session_var = datums[-1].datum.id
-        structure[session_var] = {
-            "reference": {
-                "source": "casedb",
-                "subset": "case",
-                "key": "@case_id",
+        structure["data"] = {
+            "merge": True,
+            "structure": {
+                session_var: {
+                    "reference": {
+                        "hashtag": "#case",
+                        "source": "casedb",
+                        "subset": "case",
+                        "key": "@case_id",
+                    },
+                },
+            },
+        }
+    if is_usercase_in_use(app.domain):
+        structure["context"] = {
+            "merge": True,
+            "structure": {
+                "userid": {
+                    "reference": {
+                        "hashtag": "#user",
+                        "source": "casedb",
+                        "subset": USERCASE_TYPE,
+                        "subset_key": "@case_type",
+                        "subset_filter": True,
+                        "key": "hq_user_id",
```

Comment:
The path of this `reference` is `instance('commcaresession')/session/context/userid`, and `hq_user_id` is the name of the referenced element. So this will end up creating a predicate like

`[hq_user_id = instance('commcaresession')/session/context/userid]`

The path is constructed by concatenating the instance for this source (`"id": "commcaresession"`, so `instance('commcaresession')`) with it's path (`/session`) plus the slash-delimited names of parent elements (`context`) plus the reference's element name (`userid`). Each element in the hierarchy points to a dict with keys describing it such as `structure` (sub elements), `reference` (this element references something), etc.

The thing that is referenced is the `CASE_TYPE` (`commcare-user`) subset in the `casedb` source, and since `subset_filter` is specified, an additional predicate will be added to the reference:

`[@case_type = 'commcare-user']`